### PR TITLE
Feature/git smart http

### PR DESCRIPTION
**Pull request by oliverguenther**
(This is https://github.com/opf/openproject/pull/1629, refactored and updated for dev

This PR extends `OpenProjectAuthentication.pm` to allow Git Smart HTTP of repositories from OpenProject with access through Apache, while authorization is largely identical to the subversion repositories.
### Original PR:

I'd like to see git supported in our authentication scripts (just like we support svn).

This PR aims to do that.

see: https://www.openproject.org/work_packages/14383

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-git-smart-http.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-git-smart-http.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by tessi** [comment]
Thanks @oliverguenther for keeping up with integrating this :) :clap:

**Comment by ulferts** [comment]
Hey @tessi, care to do the review? Somebody has to and you seem to be the right man for the job.

File: spec/controllers/sys_controller_spec.rb [hunk 1]
```diff
@@ -28,208 +28,549 @@
 
 require 'spec_helper'
 
-module OpenProjectRepositoryAuthenticationSpecs
```
Comment (ulferts) [inline spec/controllers/sys_controller_spec.rb:31]: You should either turn this into a request/request spec or considerably reduce the stack you are testing. You are currently not unit testing as you are testing down to the model.

I'd:
- write request specs for the happy case
- write unit test for the controller where you mock like hell
- write unit tests for the policies

That should reduce the amount of testing you are doing.

**Comment by ulferts** [comment]
Code itself is looking good but I admit to not having tested it as I have no apache set up and ready for testing.

**Comment by oliverguenther** [comment]
I have tested it myself on Apache inside a VM, however I would appreciate someone validating my findings.

I'll also take another look later today concerning the functionality after Rails 4 merge.

**Comment by tessi** [comment]
Code-wise this looks good. I don't have an apache OP setup handy. Back when I've created the first commit, the thing worked, though.

All changes by @oliverguenther since then seem to make things better. So this has my :+1:

**Comment by oliverguenther** [comment]
I've deployed this on my Apache VM and found two minor issues caused by the changes from Rails 4, but otherwise this worked fine. I'm still a little scared regarding that Git read/only check (especially regarding implementation changes in that regard), but I don't think we can make it more explicit on that side.

File: spec/controllers/sys_controller_spec.rb (hunk 1 above)
Comment (oliverguenther) [inline spec/controllers/sys_controller_spec.rb:31]: I agree. But I'd prefer to merge this and refactor the specs in a separate PR.

**Comment by tessi** [comment]
Thanks again @oliverguenther for keeping track of this. I'm happy to see this merged

**Pull request merged**